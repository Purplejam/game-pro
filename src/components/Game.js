import React from 'react';
import {motion, AnimateSharedLayout} from 'framer-motion';
import styled from 'styled-components';
import {gameDetailAction} from '../actions/gameDetailAction.js';
import {useDispatch, useSelector} from 'react-redux';
import { Link } from "react-router-dom";
import default_img from "../img/default_img.png";
import {fadeIn, item} from '../animation.js';

//Styles
const StyledGame = styled(motion.div)`
	img {
		width: 100%;
		height: 40vh;
		object-fit: cover;
	}
	h3 {
		cursor: pointer;
	}
	min-height: 30vh;
	box-shadow: 0px 5px 30px rgba(0, 0, 0, 0.2);
	text-align: center;
	border-radius: 1rem;
	overflow: hidden; 
`
//Media resize
export const smallImage = (imagePath, size) => {
  const image = imagePath.match(/media\/screenshots/)
    ? imagePath.replace(
        "media/screenshots",
        `media/resize/${size}/-/screenshots`
      )
    : imagePath.replace("/media/games/", `/media/resize/${size}/-/games/`);
  return image;
};


function Game({name, released, image, id, slug}) {
	const dispatch = useDispatch();
	const layoutId = id.toString();

	function idGameHandler() {
		document.body.style.overflow = 'hidden';
		dispatch(gameDetailAction(id, slug));
	}

	return(
		<StyledGame variants={item} onClick={idGameHandler}>
			<Link to={`/game/${id}`}>
				<motion.h3>{name}</motion.h3>	
				<p>{released}</p>
				<motion.img src={image == null ? default_img : image}/>
			</Link>	
		</StyledGame>
		)
}

export default Game;