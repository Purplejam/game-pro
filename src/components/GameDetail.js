import React from 'react';
import {motion, AnimateSharedLayout} from 'framer-motion';
import styled from 'styled-components';
import {useSelector} from 'react-redux';
import { useNavigate } from "react-router-dom";
import apple from '../img/apple.svg';
import gamepad from '../img/gamepad.svg';
import nintendo from '../img/nintendo.svg';
import playstation from '../img/playstation.svg';
import steam from '../img/steam.svg';
import xbox from '../img/xbox.svg';
import fullStar from '../img/star-full.png';
import emptyStar from '../img/star-empty.png';
import loadingGif from '../img/loader.webp';
import {gameResetAction} from '../actions/gameDetailAction.js';
import {useDispatch} from 'react-redux';

//Styles
const CardShadow = styled(motion.div)`
	position: fixed;
	overflow-y: scroll;
	min-height: 100vh;
	width: 100%;
	background-color: rgba(0,0,0,0.5);
	top: 0;
	left: 0;
	padding-top: 4rem;
	&::-webkit-scrollbar {
		width: 0.5rem;
	}
	&::-webkit-scrollbar-thumb {
		width: 0.5rem;
		background-color: #ff7676;
	}
	&::-webkit-scrollbar-track {
		width: 0.5rem;
		background-color: white;
	}
`
const Detail = styled(motion.div)`
	width: 80%;
	min-height: 90vh;
	border-radius: 0.5rem;
	padding: 2rem 5rem;
	background-color: white;
	position: absolute;
	left: 10%;
	color: black;
	img {
		width: 100%;
	}
	h3 {
		padding: 1.5rem 0rem !important;
	}
	.rating {
		text-align: left;
		img {
			display: inline;
			width: 2rem;
			height: 2rem;	
		}
	}
	.platforms {
		text-align: right;
	}
	.loading-img {
		width: 5rem;
		height: 5rem;
		margin-top: 4rem;
	}
`

const Stats = styled(motion.div)`
	display: flex;
	justify-content: space-between;
`

const Info = styled(motion.div)`
	text-align: center;
	h3 {
		text-align: right;
	}
`

const Platforms = styled(motion.div)`
	display: flex;
	justify-content: space-evenly;
	img {
		margin-left: 3rem;
	}
`

const Media = styled(motion.div)`
	margin-top: 5rem;
	img {
		width: 100%;
	}
`
const Description = styled(motion.div)`
	margin: 5rem 0rem;
`

//Get platform logo
function getGameImage(platform) {
	switch(platform) {
		case 'Apple' :
			return apple;
		case 'Xbox' : 
			return xbox;
		case 'PlayStation' : 
			return playstation;
		case 'Nintendo' : 
			return nintendo;
		case 'Steam' : 
			return steam;
		default : 
			return gamepad;
	}
}

//Get star raiting
function getStarRaiting(raiting) {
	let star = [];
	raiting = Math.floor(raiting);
	for (let i = 1; i <= 5; i++) {
		if (i <= raiting) {
			star.push(<img src={fullStar} key={i}/>)
		} else {
			star.push(<img src={emptyStar} key={i}/>)
		}
	}
	return star;
}

//Main component
function GameDetail({pathId}) {
	const {game, screen, isLoading} = useSelector(state => state.gameDetail);
	const history = useNavigate();
	const dispatch = useDispatch();

	const shadowExitHandler = (e) => {	
		if (e.target.classList.contains('shadow')) {
			dispatch(gameResetAction());
			document.body.style.overflow = 'auto';
			history('/');
		}
	}

	return(
				<CardShadow className="shadow" onClick={shadowExitHandler}>
					<Detail>
						{isLoading ? <img className="loading-img" src={loadingGif} alt="loading"/> : (
						<>
							<Stats className="stats">
								<div className="rating">
									<motion.h3>{game.name}</motion.h3>
									<p>Rating: {game.rating}</p>
									<div className="star-raiting">{getStarRaiting(game.rating)}</div>
								</div>
								<Info>
									<h3>Platforms</h3>
									<Platforms className="platforms">
										{game.parent_platforms.map(data => (
											<img key={data.platform.id} src={getGameImage(data.platform.name)}/>
											))}
									</Platforms>
								</Info>
							</Stats>
							<Media>
								<motion.img src={game.background_image} alt={game.name}/>
							</Media>
							<Description>
								<p>{game.description_raw}</p>
							</Description>
							<div className="gallery">
								{screen.results.map(screen => (
									<img src={screen.image} alt="image" key={screen.id}/>
									))}
							</div>
						</>
						)}
					</Detail>
				</CardShadow>
		)
}


export default GameDetail;