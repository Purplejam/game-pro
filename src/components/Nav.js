import {motion, AnimateSharedLayout} from 'framer-motion';
import styled from 'styled-components';
import logo from '../img/logo.svg';
import {gamesSearchThunk, clearDataAction} from '../actions/gamesAction.js';
import {useState} from 'react';
import {useDispatch} from 'react-redux';

const StyledNav = styled(motion.div)`
	padding: 3rem 5rem;
 text-align: center;
 input {
   width: 30%;
   font-size: 1.5rem;
   padding: 0.5rem;
   border: none;
   margin-top: 1rem;
   box-shadow: 0px 0px 30px rgba(0, 0, 0, 0.2);
 }
 button {
   font-size: 1.5rem;
   border: none;
   padding: 0.5rem 2rem;
   cursor: pointer;
   background: #ff7676;
   color: white;
 }
`

const Logo = styled(motion.div)`
	display: flex;
	align-items: center;
	justify-content: center;
 padding: 1rem;
 cursor: pointer;
 img {
   height: 2rem;
   width: 2rem;
 }
`

const Search = styled(motion.form)`
	
`

export default function Nav() {
	const [inputValue, setInput] = useState('');
	const dispatch = useDispatch();

	const formSubmitHandler = async (e) => {
		e.preventDefault();
		await dispatch(gamesSearchThunk(inputValue));
		setInput('');
	}

	const clearInputHandler = () => {
		dispatch(clearDataAction());
		setInput('');
	}

	return(
		<StyledNav>
			<Logo onClick={clearInputHandler}>
				<img src={logo} alt="logo"/>
				<h1>Ignite</h1>
			</Logo>
			<Search onSubmit={formSubmitHandler}>
				<input value={inputValue} type="text" onChange={(e) => setInput(e.target.value)}/>
				<button type="submit">Search</button>
			</Search>
		</StyledNav>
		)
}