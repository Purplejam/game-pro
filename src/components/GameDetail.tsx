import {motion} from 'framer-motion';
import styled from 'styled-components';
import {useSelector} from 'react-redux';
import { useNavigate } from "react-router-dom";
import apple from '../img/apple.svg';
import gamepad from '../img/gamepad.svg';
import nintendo from '../img/nintendo.svg';
import playstation from '../img/playstation.svg';
import steam from '../img/steam.svg';
import xbox from '../img/xbox.svg';
import fullStar from '../img/star-full.png';
import emptyStar from '../img/star-empty.png';
import loadingGif from '../img/loader.webp';
import {gameResetAction} from '../actions/gameDetailAction';
import {useDispatch} from 'react-redux';
import {smallImage} from './Game';
import { AppStateType } from '../reducers/index';
import { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';
import {detailReducerType, screenGameType, developerType, gamePlatFormType} from '../types';
import star from '../img/star-full.png';


//styles
const CardShadow = styled(motion.div)`
	position: fixed;
	overflow-y: scroll;
	min-height: 100vh;
	width: 100%;
	background-color: rgba(0,0,0,0.5);
	top: 0;
	left: 0;
	padding-top: 4rem;
	padding-bottom: 4rem;
	z-index: 999;
	&::-webkit-scrollbar {
		width: 0.5rem;
	}
	&::-webkit-scrollbar-thumb {
		width: 0.5rem;
		background-color: #ff7676;
	}
	&::-webkit-scrollbar-track {
		width: 0.5rem;
		background-color: white;
	}
`
const Detail = styled(motion.div)`
	width: 80%;
	min-height: 90vh;
	border-radius: 0.5rem;
	padding: 2rem 5rem;
	background-color: white;
	position: absolute;
	left: 10%;
	color: black;
	img {
		width: 100%;
	}
	.gallery {
		img {
			margin-bottom: 0.5rem;
		}
	}
	h3 {
		padding: 1.5rem 0rem;
	}
	.rating {
		text-align: left;
		width: 50%;
		img {
			display: inline;
			width: 2rem;
			height: 2rem;	
		}
	}
	.platforms {
		text-align: right;
	}
	.game-website {
		padding-top: 4rem;
		display: flex;
		justify-content: space-between;
		flex-direction: row;
		img {
			width: 1.2rem;
			height: 1.2rem;
			display: inline-block;
			margin-right: 0.5rem;
			margin-top: -0.5rem;
		}
	}
	.loading-img {
		width: 5rem;
		height: 5rem;
		margin-top: 4rem;
	}
	 @media (max-width: 768px) {
	 	width: 96%;
	 	left: 2%;
	 	padding: 1rem 1rem;
	 	.game-website {
	 		img {
			 	width: 0.8rem;
					height: 0.8rem;
					margin-right: 0.2rem;
	 		}
	 	}
 }
`

const Stats = styled(motion.div)`
	display: flex;
	justify-content: space-between;
	@media (max-width: 768px) {
		.star-raiting {
			img {
				width: 1rem;
				height: 1rem;
			}
		}
 }
`

const Info = styled(motion.div)`
	text-align: center;
	width: 50%;
	h3 {
		text-align: right;
	}
`

const Platforms = styled(motion.div)`
	display: flex;
	flex-direction: row;
	justify-content: flex-end;
	align-content: center;
	img {
		display: inline-block;
		margin-left: 1rem;
		width: 2rem;
	}
	@media (max-width: 768px) {
		img {
			width: 1.5rem;
		}
	}
	@media (max-width: 400px) {
		img {
			width: 1.2rem;
		}
	}
`

const Media = styled(motion.div)`
	margin-top: 5rem;
	img {
		width: 100%;
	}
	@media (max-width: 768px) {
		margin-top: 3rem;
	}
	@media (max-width: 400px) {
		margin-top: 1.5srem;
	}
`
const Description = styled(motion.div)`
	margin: 5rem 0rem;
	@media (max-width: 768px) {
		margin: 3rem 0rem;
	}
	@media (max-width: 400px) {
		margin: 1.5rem 0rem;
	}
`
const Developers = styled(motion.div)`
	.developers-item {
		display: flex;
		flex-direction: row;
		align-items: center;
		text-align: left;
		padding: 0.5rem 0rem;
		img {
			width: 30%;
			border-radius: 5px;
		}
		.developers-description {
			p {
				margin: 0rem 0rem;
				padding: 0rem 0rem;
			}
			padding: 0rem 1rem;
		}
	}
`

//get platform logo
function getGameImage(platform: string | null) {
	switch(platform) {
		case 'Apple' :
			return apple;
		case 'Xbox' : 
			return xbox;
		case 'PlayStation' : 
			return playstation;
		case 'Nintendo' : 
			return nintendo;
		case 'Steam' : 
			return steam;
		default : 
			return gamepad;
	}
}

//get star raiting
function getStarRaiting(raiting: number | null) {
	let star = [];
	if (raiting !== null) {
		raiting = Math.floor(raiting);
		for (let i = 1; i <= 5; i++) {
			if (i <= raiting) {
				star.push(<img src={fullStar} key={i} alt="Full Star"/>)
			} else {
				star.push(<img src={emptyStar} key={i} alt="Empty Star"/>)
			}
		}
		return star;
	}
}

//components wrapper
function GameDetailWrapper() {
	const {game, screen, isLoading} = useSelector((state: AppStateType) => state.gameDetail);
	return(
		<GameDetail game={game} screen={screen} isLoading={isLoading}/>
		)
}

//main component
export function GameDetail({game, screen, isLoading}: detailReducerType) {
	
	const history = useNavigate();
	const dispatch: ThunkDispatch<AppStateType, void, Action> = useDispatch();

	const shadowExitHandler = (e: React.MouseEvent<HTMLDivElement>) => {	
		const target = e.target as Element;
		if (target.classList.contains('shadow')) {
			dispatch(gameResetAction());
			document.body.style.overflow = 'auto';
			history('/');
		}
	}

	return(
				<CardShadow className="shadow" onClick={shadowExitHandler}>
					<Detail>
						{isLoading ? <img className="loading-img" src={loadingGif} alt="loading"/> : (
						<>
							<Stats className="stats">
								<div className="rating">
									<motion.h3>{game.name}</motion.h3>
									<p>Rating: {game.rating}</p>
									<div className="star-raiting">{getStarRaiting(game!.rating)}</div>
								</div>
								<Info>
									<h3>Platforms</h3>
									<Platforms className="platforms">
										{game.parent_platforms.map((data: gamePlatFormType) => (
											<img alt={data.platform.name} key={data.platform.id} src={getGameImage(data.platform.name)}/>
											))}
									</Platforms>
								</Info>
							</Stats>
							<Media>
								<motion.img alt={game.name} src={smallImage(game.background_image, 1280)}/>
							</Media>
							<Description>
								<p>{game.description_raw}</p>
							</Description>
							<div className="gallery">
							<h3>Screenshots</h3>
								{screen.results.map((screen: screenGameType) => (
									<img src={smallImage(screen.image, 1280)} alt={`${screen.id}`} key={screen.id}/>
									))}
							</div>
								{game.reddit_description.length > 0 && (
									<Description>
										<h3>Reddit description</h3>
										<p>{game.reddit_description}</p>
									</Description>)}
								{game.developers.length > 0 && (
									<Developers>
										<h3>Developers</h3>
										{game.developers.map((dev: developerType) => (
											<div className="developers-item" key={dev.id}>
												<img src={smallImage(dev.image_background, 1280)} alt={`${dev.id}`} key={dev.id}/>
												<div className="developers-description">
													<p><b>Developer: </b>{dev.name}</p>
													<p><b>{`${dev.games_count}`}</b> published games</p>
												</div>
											</div>
											))}
									</Developers>
									)}
								<div className="game-website">
									<p><img src={star} alt={game.name}/><b>{game.added}</b></p>
									<p><a href={game.website}>{game.website}</a></p>
								</div>
						</>
						)}
					</Detail>
				</CardShadow>
		)
}


export default GameDetailWrapper;