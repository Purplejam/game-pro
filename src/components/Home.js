import React, {useEffect, useState} from 'react';
import ReactDOM from 'react-dom/client';
import {useDispatch, useSelector} from 'react-redux';
import {gamesActionThunk} from '../actions/gamesAction.js';
import Game from './Game';
import {motion} from 'framer-motion';
import styled from 'styled-components';
import {gameDetailsURL} from '../api.js';
import axios from "axios";
import { useLocation } from "react-router-dom";
import GameDetail from './GameDetail.js';
import Nav from './Nav.js';
import loadingGif from '../img/loader.webp';
import {fadeIn, item} from '../animation.js';


//Styles
const GameList = styled(motion.div)`
	padding: 0rem 5rem 5rem 5rem;
	text-align: center;
		h4 {
			padding: 5rem 0rem 4rem 0rem;
		}
	.game-categories {
		display: grid;
		grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
			h3 {
				cursor: pointer !important;
				padding: 5rem 0rem 4rem 0rem;
			}
		.active {
			color: #ff7676;;
		}
		.deactive {
			color: #333;
		}
	}
	.loading-img {
		margin: 0 auto;
		opacity: 0.9;
	}
`
const Games = styled(motion.div)`
	min-height: 80vh;
	display: grid;
	grid-template-columns: repeat(auto-fit, minmax(500px, 1fr));
	grid-column-gap: 3rem;
	grid-row-gap: 5rem;
`

//Tabs component
const CategoryHeader = ({name, setActiveCategoryHandler, activeNav}) => {
	return(
		<h3 className={`${activeNav == name ? "active" : "deactive"}`}
			onClick={() => setActiveCategoryHandler(name)}>
			{name}
		</h3>
		)
}
	
//Main component
function Home() {
	const location = useLocation();
	const path = location.pathname.split('/');
	const pathId = path[path.length - 1];
	const dispatch = useDispatch();
	const {popular, upcoming, newComes, isLoading, searched} = useSelector((state) => state.games);
	
	useEffect(() => {
		dispatch(gamesActionThunk());
	}, [dispatch])

	useEffect(() => {
		setActiveList(popular);
	}, [popular])

	const [activeList, setActiveList] = useState([]);
	const [activeNav, setActiveNav] = useState('Popular Games');

	const setActiveCategoryHandler = (name) => {
		switch(name) {
			case 'Newest Games' : 
			 setActiveList(newComes);
			 setActiveNav(name);
			 break;
			case 'Popular Games' : 
			 setActiveList(popular);
			 setActiveNav(name);
			 break;
			case 'Upcoming Games' : 
			 setActiveList(upcoming);
			 setActiveNav(name);
			 break;
			default :
				setActiveList(popular);
				setActiveNav(popular);
		}
	}

	return (	
		<GameList>
			<Nav/>
			{isLoading ? <img className="loading-img" src={loadingGif} alt="loading"/> : (<>
			{(searched.games.length > 0) && <div className="searched-items">
				<h4>{`We found ${searched.games.length} items of "${searched.query}":`}</h4>
				<Games variants={fadeIn} initial="hidden" animate="show">
				{searched.games.map(game => {
					return <Game  
					slug={game.slug}
					name={game.name} 
					id={game.id} 
					key={game.id} 
					released={game.released} 
					image={game.background_image}/>
					})}
				</Games>
			</div> }
			{pathId && <GameDetail layout/>}
			<div className="game-categories">			
				<CategoryHeader name="Popular Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
				<CategoryHeader name="Upcoming Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
				<CategoryHeader name="Newest Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
			</div>
				<Games variants={fadeIn} initial="hidden" animate="show">
				{activeList.map(game => {
					return <Game					
					slug={game.slug}
					name={game.name} 
					id={game.id} 
					key={game.id} 
					released={game.released} 
					image={game.background_image}/>
					})}
			</Games>
			</>)}
		</GameList>
		)
}


export default Home;