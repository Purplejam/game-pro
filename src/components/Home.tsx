import {useEffect, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {gamesActionThunk} from '../actions/gamesAction';
import Game from './Game';
import {motion} from 'framer-motion';
import styled from 'styled-components';
import { useLocation } from "react-router-dom";
import GameDetailWrapper from './GameDetail';
import Nav from './Nav';
import loadingGif from '../img/loader.webp';
import {fadeIn} from '../animation';
import { AppStateType } from '../reducers/index';
import { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';

//types
export type gameType = {
	id: number,
	name: string,
	ratings: {
		percent: number
	}[],			
	slug: string,
	added: number,
	released: string,
	background_image: string,
	genres: {
		name: string,
		id: number
	}[]
}

export type gameListType = Array<gameType>;

//styles
const GameList = styled(motion.div)`
	padding: 0rem 5rem 5rem 5rem;
	text-align: center;
		h4 {
			padding: 5rem 0rem 5rem 0rem;
		}
	.game-categories {
		display: grid;
		grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
		padding: 2rem 0rem;
			h3 {
				cursor: pointer !important;
				padding: 2rem 0rem 1rem 0rem;
				position: relative;
				&::before {
				  content: "";
				  position: absolute;
				  display: block;
				  width: 100%;
				  height: 2px;
				  bottom: 0;
				  background-color: #ff7676;
				  transform: scaleX(0);
				  transition: transform 0.3s ease;
				  left: 0;
				}
				&:hover::before {
				  transform: scaleX(1);
				}
			}
		.active {
			color: #ff7676;
		}
		.deactive {
			color: #333;
		}
	}
	.loading-img {
		margin: 0 auto;
		opacity: 0.9;
	}
	@media (max-width: 768px) {
		padding: 0rem 2rem 2rem 2rem;
		.game-categories {
			grid-template-columns: repeat(auto-fit, minmax(320px, 1fr));
			h3 {
				padding: 1rem 0rem 1rem 0rem;
			}
		}
 }
	@media (max-width: 400px) {
		padding: 0rem 1rem 1rem 1rem;
 }
`
const Games = styled(motion.div)`
	min-height: 80vh;
	display: grid;
	grid-template-columns: repeat(auto-fit, minmax(360px, 1fr));
	grid-column-gap: 3rem;
	grid-row-gap: 5rem;
	@media (max-width: 768px) {
		grid-template-columns: repeat(auto-fit, minmax(300px, 1fr));
		grid-row-gap: 2rem;
 }
`

type CategoryHeaderType = {
	name: string,
	setActiveCategoryHandler: Function,
	activeNav: string
}

//tabs component
const CategoryHeader = ({name, setActiveCategoryHandler, activeNav}: CategoryHeaderType) => {
	return(
		<h3 className={`${activeNav === name ? "active" : "deactive"}`}
			onClick={() => setActiveCategoryHandler(name)}>
			{name}
		</h3>
		)
}
	
//main component
function Home() {
	const location = useLocation();
	const path = location.pathname.split('/');
	const pathId: string = path[path.length - 1];
	const dispatch: ThunkDispatch<AppStateType, void, Action> = useDispatch();
	const {popular, upcoming, newComes, isLoading, searched} = useSelector((state: AppStateType) => state.games);
	
	useEffect(() => {
		dispatch(gamesActionThunk());
	}, [dispatch])

	useEffect(() => {
		setActiveList(upcoming);
	}, [upcoming])

	const [activeList, setActiveList] = useState<gameListType>();
	const [activeNav, setActiveNav] = useState('Upcoming Games');

	const setActiveCategoryHandler = (name: string) => {
		switch(name) {
			case 'Newest Games' : 
			 setActiveList(newComes);
			 setActiveNav(name);
			 break;
			case 'Popular Games' : 
			 setActiveList(popular);
			 setActiveNav(name);
			 break;
			case 'Upcoming Games' : 
			 setActiveList(upcoming);
			 setActiveNav(name);
			 break;
			default :
				setActiveList(upcoming);
				setActiveNav('Upcoming Games');
		}
	}

	return (	
		<GameList>
			<Nav/>
			{isLoading ? <img className="loading-img" src={loadingGif} alt="loading"/> : (<>
			{(searched.games.length > 0) && <div className="searched-items">
				<h4>{`We found ${searched.games.length} items of "${searched.query}":`}</h4>
				<Games variants={fadeIn} initial="hidden" animate="show">
				{searched.games.map((game: gameType) => {
					return <Game  
					slug={game.slug}
					name={game.name} 
					id={game.id} 
					key={game.id} 
					released={game.released} 
					background_image={game.background_image}
					genres={game.genres}
					ratings={game.ratings}
					added={game.added}/>
					})}
				</Games>
			</div> }
			{pathId && <GameDetailWrapper/>}
			<div className="game-categories">			
				<CategoryHeader name="Popular Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
				<CategoryHeader name="Upcoming Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
				<CategoryHeader name="Newest Games" activeNav={activeNav} setActiveCategoryHandler={setActiveCategoryHandler}/>
			</div>
				<Games variants={fadeIn} initial="hidden" animate="show">
				{activeList?.map(game => {
					return <Game		
					ratings={game.ratings}			
					slug={game.slug}
					added={game.added}
					name={game.name} 
					id={game.id} 
					key={game.id} 
					released={game.released} 
					background_image={game.background_image}
					genres={game.genres}/>
					})}
			</Games>
			</>)}
		</GameList>
		)
}


export default Home;