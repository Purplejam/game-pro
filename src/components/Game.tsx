import {motion} from 'framer-motion';
import styled from 'styled-components';
import {gameDetailAction} from '../actions/gameDetailAction';
import {useDispatch} from 'react-redux';
import { Link } from "react-router-dom";
import default_img from "../img/default_img.png";
import { item } from '../animation';
import star from '../img/star-full.png';
import { gameType } from './Home';
import { AppStateType } from '../reducers/index';
import { ThunkDispatch } from 'redux-thunk';
import { Action } from 'redux';


//styles
const StyledGame = styled(motion.div)`
	img {
		width: 100%;
		height: 40vh;
		object-fit: cover;
	}
	h3 {
		cursor: pointer;
		padding: 2rem 0rem 0.5rem 0rem;
	}
	min-height: 30vh;
	box-shadow: 0px 5px 30px rgba(0, 0, 0, 0.2);
	text-align: center;
	border-radius: 1rem;
	overflow: hidden; 
	.game-meta-info {
		display: flex;
		flex-direction: row;
		justify-content: space-between;
		align-content: space-between;
		padding: 1rem 2rem;
			img {
				width: 1.2rem;
				height: 1.2rem;
				display: inline-block;
				margin-left: 0.5rem;
				margin-top: -0.1rem;
				vertical-align: middle;
			}
			div, p {
				display: inline-block;
				font-weight: 700;
			}
			.game-raiting {
				display: flex;
				align-items: center;
			}
	}
	.game-genres {
		text-align: center;
		padding: 1rem 2rem;
	}
 @media (max-width: 768px) {
 }
`
//media resize
export const smallImage = (imagePath: string, size: number) => {
  const image = imagePath.match(/media\/screenshots/)
    ? imagePath.replace(
        "media/screenshots",
        `media/resize/${size}/-/screenshots`
      )
    : imagePath.replace("/media/games/", `/media/resize/${size}/-/games/`);
  return image;
};

//rating color
const ratingColor = (rate: number) => {
	if (rate >= 70 && rate < 100) {
		return 'green';
	} else if (rate >= 50 && rate < 70) {
		return 'orange';
	} else if (rate > 0 && rate < 50) {
		return 'red';
	}
}


function Game({name, released, background_image, id, slug, genres, ratings, added}: gameType) {
	const dispatch: ThunkDispatch<AppStateType, void, Action> = useDispatch();

	function idGameHandler() {
		document.body.style.overflow = 'hidden';
		dispatch(gameDetailAction(id, slug));
	}

	return(
		//@ts-ignore
		<StyledGame variants={item} onClick={idGameHandler}>
			<Link to={`/game/${id}`}>
				<motion.h3>{name}</motion.h3>	
				<p>Released: {released}</p>
				<motion.img src={background_image == null ? default_img : smallImage(background_image, 420)}/>
				{(genres.length > 0) 
					? <p className="game-genres">Genre{genres.length > 1 ? 's' : null}: {genres.map(genre => <span key={genre.id}>{genre.name + ' '}</span>)}</p>
					: <p className="game-genres">Genre: video game</p>}
				<div className="game-meta-info">
					{ratings.length !== 0 ? <p style={{color: ratingColor(ratings[0].percent)}}>Raiting: {Math.floor(ratings[0].percent)}</p> : null}
					<div className="game-raiting">
						<p>{added}</p>
						<img src={star} alt=""/>
					</div>
				</div>
			</Link>	
		</StyledGame>
		)
}

export default Game;