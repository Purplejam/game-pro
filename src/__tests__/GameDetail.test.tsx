import {render, screen} from '@testing-library/react';
import {GameDetail} from '../components/GameDetail';
import {store} from '../store/store';
import { Provider } from 'react-redux';
import React from 'react';
import { BrowserRouter} from "react-router-dom";
import '@testing-library/jest-dom';

 
const gameDetailMock = {
	game: {
		name: 'TopGame',
		id: 1234,
		rating: 44,
		parent_platforms: [{
			platform: {
				id: 1,
				name: 'Apple'
			}
		}],
		background_image: 'game-img.jpg',
		description_raw: 'lorem ipsum'
	},
	screen: {
		results: [{
					id: 1234,
					image: 'game-img.jpg'
				}]
	},
	isLoading: false,
}

const gameDetailMockLoading = {
	...gameDetailMock,
	isLoading: true,
}

describe('<GameDetail/>', () => {
	it('renders properly the game component', () => {
		const {container} = render(
			<BrowserRouter>
				<React.StrictMode>
					<Provider store={store}>
						 <GameDetail {...gameDetailMock}/>
					</Provider>
				</React.StrictMode>
			</BrowserRouter>
			)

		const nameText = screen.getByText('TopGame');
		expect(nameText).toBeInTheDocument();

		const platformText = screen.getByAltText('Apple');
		expect(platformText).toBeInTheDocument();

		expect(container).toMatchSnapshot();
	})

		it('renders properly while loading', () => {
		render(
			<BrowserRouter>
				<React.StrictMode>
					<Provider store={store}>
						 <GameDetail {...gameDetailMockLoading}/>
					</Provider>
				</React.StrictMode>
			</BrowserRouter>
			)

		const platformText = screen.getByAltText('loading');
		expect(platformText).toBeInTheDocument();

	})
})

