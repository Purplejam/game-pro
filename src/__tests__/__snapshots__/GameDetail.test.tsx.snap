// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameDetail/> renders properly the game component 1`] = `
<div>
  <div
    class="sc-gsnTZi kCCLP shadow"
  >
    <div
      class="sc-dkzDqf cdpgAH"
    >
      <div
        class="sc-hKMtZM eHEAZC stats"
      >
        <div
          class="rating"
        >
          <h3>
            TopGame
          </h3>
          <p>
            Rating: 
            44
          </p>
          <div
            class="star-raiting"
          >
            <img
              alt="Full Star"
              src="star-full.png"
            />
            <img
              alt="Full Star"
              src="star-full.png"
            />
            <img
              alt="Full Star"
              src="star-full.png"
            />
            <img
              alt="Full Star"
              src="star-full.png"
            />
            <img
              alt="Full Star"
              src="star-full.png"
            />
          </div>
        </div>
        <div
          class="sc-eCYdqJ hyBjDA"
        >
          <h3>
            Platforms
          </h3>
          <div
            class="sc-jSMfEi hfCYlh platforms"
          >
            <img
              alt="Apple"
              src="apple.svg"
            />
          </div>
        </div>
      </div>
      <div
        class="sc-gKXOVf kVtWqt"
      >
        <img
          alt="TopGame"
          src="game-img.jpg"
        />
      </div>
      <div
        class="sc-iBkjds bKRqVj"
      >
        <p>
          lorem ipsum
        </p>
      </div>
      <div
        class="gallery"
      >
        <img
          alt="1234"
          src="game-img.jpg"
        />
      </div>
    </div>
  </div>
</div>
`;
